library/std/src/thread/scoped.rs

```diff
@@ -35,28 +38,148 @@ pub struct Scope<'scope, 'env: 'scope> {
 #[stable(feature = "scoped_threads", since = "1.63.0")]
 pub struct ScopedJoinHandle<'scope, T>(JoinInner<'scope, T>);
 
+/// Set by the main thread to indicate that `main_thread` has a value
+/// and it is waiting to be notified by zeroing out the `sync_state`.
+const WAITING_BIT: usize = 1;
+
+/// Represents one thread spawned by the scope that is actively running.
+const ONE_RUNNING: usize = 2;
+
+/// A soft limit on the amount of concurrent spawned threads in a scope.
+///
+/// Going above this limit will panic to avoid the `sync_state` wrapping
+/// around and potetially triggering undefined behavior.
+const MAX_RUNNING: usize = usize::MAX / 2;
+
+#[derive(Default)]
 pub(super) struct ScopeData {
-    num_running_threads: AtomicUsize,
+    sync_state: AtomicUsize,
     a_thread_panicked: AtomicBool,
-    main_thread: Thread,
+    main_thread: Cell<Option<Thread>>,
+    _pinned: PhantomPinned,
 }
 
+unsafe impl Send for ScopeData {}
+unsafe impl Sync for ScopeData {}
```

Comment:
Using `Cell` together with `unsafe impl Sync` is probably a bad idea. It removes all checks: any thread can now just call `.get()` and `.set()` on it, without any `unsafe` block.

Reply:
Sure. I'm not saying that your code is unsound. I'm saying that it's very easy to make a mistake when forcing a Cell into a Sync type. We usually avoid that and keep using UnsafeCell to make sure we have to write `unsafe` when using it, so it can't accidentally be used from the wrong thread without thinking.

Reply:
Oh, that makes sense.